python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,191 @@
+    ):
+        # When the last remaining reference is being cleaned up we will still
+        # have a dead weak-reference in `weak_set_values`, if that is the case
+        # we are good to perform the key's cleanup
+        del CopyOnWriteBuffer._instances[(ptr, size)]
+
+
+class CopyOnWriteBuffer(Buffer):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    # This dict keeps track of all instances that have the same `ptr`
+    # and `size` attributes.  Each key of the dict is a `(ptr, size)`
+    # tuple and the corresponding value is a set of weak references to
+    # instances with that `ptr` and `size`.
+    _instances: DefaultDict[Tuple, WeakSet] = defaultdict(WeakSet)
+
+    # TODO: This is synonymous to SpillableBuffer._exposed attribute
+    # and has to be merged.
+    _zero_copied: bool
+
+    def _finalize_init(self):
+        key = (self._ptr, self._size)
+        self.__class__._instances[key].add(self)
+        self._zero_copied = False
+        weakref.finalize(self, _keys_cleanup, self._ptr, self._size)
+
+    @classmethod
+    def _from_device_memory(
+        cls: Type[T], data: Any, *, exposed: bool = False
+    ) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        ret._zero_copied = exposed
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @property
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self.__class__._instances[(self._ptr, self._size)]) > 1
+
+    @property
+    def ptr(self) -> int:
+        """Device pointer to the start of the buffer."""
+        self._unlink_shared_buffers()
+        self._zero_copied = True
+        return self._ptr
+
+    @property
+    def mutable_ptr(self) -> int:
+        """Device pointer to the start of the buffer."""
+        self._unlink_shared_buffers()
+        return self._ptr
+
+    def _getitem(self, offset: int, size: int) -> Buffer:
+        """
+        Helper for `__getitem__`
+
+        Returns the same underlying memory pointer if offset is 0
+        and size == self.size, else makes a copy to return the
+        slice.
+        """
+        if offset != 0 or self.size != size:
+            buf = self.copy(deep=True)
```

Comment:
So the reason for us to perform an eager copy is to ensure we don't have overlapping Buffers in our WeakSet. @vyasr brought this up here https://github.com/rapidsai/cudf/pull/11718#discussion_r1062692454 and I realized `Buffer._getitem` was marking the `CopyOnWriteBuffer` as zero copied(which is incorrect) and  enforcing a new copy when `.ptr` was being accessed in `Buffer._getitem`. `CopyOnWriteBuffer._getitem` bypasses marking the `CopyOnWriteBuffer` as zero copied and creates a copy only if a slice is asked for and returns the same memory if the entire buffer is asked for.

Reply:
@galipremsagar I assume this is the fix that you were alluding to in https://github.com/rapidsai/cudf/pull/11718#discussion_r1065142419? I didn't appreciate how serious an issue this could be when I raised the question, but Lawrence is right. This potentially makes COW _much_ more expensive.

Fixing this seems a bit tricky, but perhaps not too bad (hopefully, let's brainstorm). Currently the key is always `(self._ptr, self._size)`, and it is always constructed on demand. What if we stored the key as `self._key`, then for slices we instead store the key as the ptr and size of the original buffer being sliced? Then you would have the following semantics:
```
>>> x = cudf.Series([1, 2, 3]) # ptr = x, size = 24, key = (x, 24)
>>> y = x[1:] # ptr = x + 8, size = 16, key = (x, 24)
>>> x[:] = 10 # this triggers a copy because y and x have the same key
```

With this change, I'm not sure we need the size to be part of the key. Would it be sufficient to key the dict of instances on the ptr alone?